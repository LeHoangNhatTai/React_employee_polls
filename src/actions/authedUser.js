export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_USER = "LOGOUT_USER";

export function setAuthedUser(id) {
  return {
    type: SET_AUTHED_USER,
    id,
  };
}

export function logout(){
  return {
    type: LOGOUT_USER,
  }
}

export function handleLogin(userName, password) {
  return (dispatch, getState) => {
    const { users } = getState();
    const user = Object.values(users).find(user => user.id === userName && user.password === password);

    if (user) {
      dispatch(setAuthedUser(user.id));
    }
    else {
      alert('Invalid username or password');
    }
  }
}

export function handleLogout() {
  return dispatch => dispatch(logout());
}