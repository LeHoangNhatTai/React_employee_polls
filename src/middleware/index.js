import { thunk } from "redux-thunk";
import logger from "./logger";

const customMiddleware = [thunk, logger];

export default customMiddleware;