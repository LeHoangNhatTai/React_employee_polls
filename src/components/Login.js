import React, { useState } from "react";
import { connect } from "react-redux";
import { handleLogin } from "../actions/authedUser";

const Login = ({ dispatch }) => {
    const [userName, setUserName] = useState("");
    const [password, setPassword] = useState("");

    const handleSubmit = (e) => {
        e.preventDefault();
        dispatch(handleLogin(userName, password));
        setPassword("");
    }

    return (
        <div className="container login-form">
            <h2 className="center">Employee Polls</h2>
            <div className="box-center">
                <div className="login-avatar"></div>
            </div>
            <h3 className="center" data-testid="login-header">Login</h3>
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    id="userName"
                    placeholder="Username"
                    name="userName"
                    value={userName}
                    onChange={(e) => setUserName(e.target.value)}
                    data-testid="username"
                />
                <input
                    type="password"
                    id="password"
                    placeholder="Password"
                    name="password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                    data-testid="password"
                />
                <div className="box-center">
                    <button type="submit" className="btn center" data-testid="submit">Submit</button>
                </div>
            </form>
        </div>
    )
}

const mapStateToProps = ({ users }) => ({ users });

export default connect(mapStateToProps)(Login);
