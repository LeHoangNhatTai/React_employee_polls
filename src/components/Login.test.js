import '@testing-library/jest-dom';
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { legacy_createStore as createStore } from 'redux';
import Login from "./Login";
import reducer from "../reducers";
import middleware from "../middleware";
import { handleInitialData } from "../actions/shared";

describe("Login", () => {
    it("render the component", () => {
        const store = createStore(reducer);

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it('show input element', async () => {
        const store = createStore(reducer);
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        const usernameInputElement = component.getByTestId("username");
        const passwordInputElement = component.getByTestId("password");
        expect(usernameInputElement).toBeInTheDocument();
        expect(passwordInputElement).toBeInTheDocument();
    });

    it("valid process submit button", async () => {
        const store = createStore(reducer,middleware);
        await store.dispatch(handleInitialData());
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        const usernameInputElement = component.getByTestId("username");
        const passwordInputElement = component.getByTestId("password");
        const buttonSubmitElement = component.getByTestId("submit");
        expect(usernameInputElement).toBeInTheDocument();
        expect(passwordInputElement).toBeInTheDocument();
        expect(buttonSubmitElement).toBeInTheDocument();

        fireEvent.change(usernameInputElement, { target: { value: 'zoshikanlu' } });
        fireEvent.change(passwordInputElement, { target: { value: 'pass246' } });
        expect(usernameInputElement.value).toBe("zoshikanlu");
        expect(passwordInputElement.value).toBe("pass246");
        expect(buttonSubmitElement).toBeEnabled();
        fireEvent.click(buttonSubmitElement);
    });
});