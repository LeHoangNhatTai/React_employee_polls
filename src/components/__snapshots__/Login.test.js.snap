// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container login-form"
      >
        <h2
          class="center"
        >
          Employee Polls
        </h2>
        <div
          class="box-center"
        >
          <div
            class="login-avatar"
          />
        </div>
        <h3
          class="center"
          data-testid="login-header"
        >
          Login
        </h3>
        <form>
          <input
            data-testid="username"
            id="userName"
            name="userName"
            placeholder="Username"
            type="text"
            value=""
          />
          <input
            data-testid="password"
            id="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <div
            class="box-center"
          >
            <button
              class="btn center"
              data-testid="submit"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container login-form"
    >
      <h2
        class="center"
      >
        Employee Polls
      </h2>
      <div
        class="box-center"
      >
        <div
          class="login-avatar"
        />
      </div>
      <h3
        class="center"
        data-testid="login-header"
      >
        Login
      </h3>
      <form>
        <input
          data-testid="username"
          id="userName"
          name="userName"
          placeholder="Username"
          type="text"
          value=""
        />
        <input
          data-testid="password"
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <div
          class="box-center"
        >
          <button
            class="btn center"
            data-testid="submit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
